File: 082__Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head

        prev, curr = dummy, head
        while curr:
            check = False
            while curr and curr.next and curr.val == curr.next.val:
                check = True
                curr.next = curr.next.next
            if check:
                prev.next = curr.next
                curr = curr.next
            else:
                prev, curr = curr, curr.next
        return dummy.next

File: Java/082__Remove_Duplicates_from_Sorted_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode cur = head, ptr = head.next;
        ListNode prev = new ListNode(-1);
        prev.next = head;
        while (cur != null) {
            if (cur.next != null && cur.val == cur.next.val) {
                while (ptr != null && ptr.val == cur.val) {
                    ptr = ptr.next;
                }
                if (cur == head) {
                    head = ptr;
                }
                prev.next = ptr;
                cur = ptr;
            } else {
                prev = cur;
                cur = cur.next;
                ptr = cur;
            }
        }
        return head;
    }
}
